test/boost/cache_algorithm_test.cc

```diff
@@ -0,0 +1,152 @@
+
+
+        constexpr uint64_t hot_subset_size = 1000;
+        // Sanity check. The test assumes that the total *hot* index size is significantly bigger than RAM.
+        BOOST_REQUIRE_GT(hot_subset_size * pk_size * 2000, 2 * seastar::memory::stats().total_memory());
+
+        auto get_misses = [&e] { return e.local_db().row_cache_tracker().get_stats().partition_misses; };
+        uint64_t misses_before = get_misses();
+        for (size_t i = 0; i < hot_subset_size; ++i) {
+            e.execute_prepared(select_query, {{cql3::raw_value::make_value(make_key(i))}}).get0();
+            // Sanity check. If a single query is causing multiple partition misses,
+            // something unexpected is happening.
+            BOOST_REQUIRE_LE(get_misses() - misses_before, i + 1);
+        }
+
+        misses_before = get_misses();
+        // The rows we just read have a small total size. They should be perfectly cached.
+        for (size_t repeat = 0; repeat < 3; ++repeat) {
+            for (size_t i = 0; i < hot_subset_size; ++i) {
+                e.execute_prepared(select_query, {{cql3::raw_value::make_value(make_key(i))}}).get0();
+                // If the rows were perfectly cached, there were no new misses.
+                BOOST_REQUIRE_EQUAL(get_misses(), misses_before);
+            }
+        }
+    }, std::move(cfg));
+}
+
+// The previous test checks that index_cache_fraction doesn't allow index cache to
+// flood the memory.
+//
+// This test checks that it doesn't completely kill caching.
+SEASTAR_TEST_CASE(test_index_is_cached_in_big_partition_workload) {
+    cql_test_config cfg;
+    cfg.db_config->task_ttl_seconds.set(0);
+    // As of this writing, uncommenting the below should make the test fail.
+    // cfg.db_config->index_cache_fraction.set(0.0);
+    return do_with_cql_env_thread([] (cql_test_env& e) {
+        // We disable compactions because they cause confusing cache mispopulations.
+        e.execute_cql("CREATE TABLE ks.t(pk bigint, ck bigint, v blob, primary key (pk, ck)) WITH compaction = { 'class' : 'NullCompactionStrategy' };").get();
+        auto insert_query = e.prepare("INSERT INTO ks.t(pk, ck, v) VALUES (?, ?, ?)").get0();
+        auto select_query = e.prepare("SELECT * FROM t WHERE pk = ? AND ck = ?").get0();
+
+        constexpr uint64_t pk_number = 10;
+        constexpr uint64_t ck_number = 300;
+        constexpr uint64_t v_size = 100000;
+        // Sanity check. The test assumes that the total table size is significantly bigger than RAM.
+        BOOST_REQUIRE_GT(pk_number * ck_number * v_size, 2 * seastar::memory::stats().total_memory());
+
+        // A bijection between uint64_t and blobs of size pk_size.
+        auto make_key = [] (uint64_t x) {
+            bytes b(bytes::initialized_later(), sizeof(x));
+            auto i = b.begin();
+            write<uint64_t>(i, x);
+            return b;
+        };
+
+        // Populate the table.
+        auto pk = bytes(sizeof(uint64_t), 0);
```

Comment:
pk is unused?

Reply:
Right, this line is leftover from an earlier iteration of the test.